Hw1

6ECCMEPTHIoIIJ opened the pull request.

6ECCMEPTHIoIIJ committed:
Task0 made function
Task1 made function & tests
Task2 made function
Task3 made function

6ECCMEPTHIoIIJ committed:
Cleanup code
Task2 made tests

6ECCMEPTHIoIIJ committed:
Task0-Task8 functions added

6ECCMEPTHIoIIJ committed:
Task4-Task7 tests added

6ECCMEPTHIoIIJ committed:
Task8 test added

vasilmik commented on src/main/java/edu/hw1/Task1.java:
```diff
@@ -0,0 +1,33 @@
+    private final static String TIME_STRING_REGEX = "\\d{2,}:[0-5][0-9]";
+    private final static Pattern TIME_PATTERN = Pattern.compile(TIME_STRING_REGEX);
+
+    private Task1() {
```
Удалите не используемые методы

6ECCMEPTHIoIIJ replied:
private Task1() - приватный конструктор утилитного класса. Если его удалить, то код не пройдет checkstyle.

vasilmik commented on src/main/java/edu/hw1/Task1.java:
```diff
@@ -0,0 +1,33 @@
+    public static long minutesToSeconds(String timeString)
+        throws NullPointerException {
+        if (timeString == null) {
+            throw new NullPointerException();
```
Более подходит другое исключение

vasilmik commented on src/test/java/edu/hw1/Task2Test.java:
```diff
@@ -0,0 +1,48 @@
+import org.junit.jupiter.params.provider.MethodSource;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+public class Task2Test {
```
Проверьте  corner case и увидите что алгоритм не всегда работает верно

6ECCMEPTHIoIIJ replied:
Действительно, тест упал, когда проверил на Integer.MIN_VALUE из-за переполнения типа при взятии модуля. По итогу возвращалось некорректное значение 1. 
Сделал отрицательные числа не валидными

vasilmik commented on src/main/java/edu/hw1/Task6.java:
```diff
@@ -0,0 +1,77 @@
+
+    }
+
+    private static boolean hasNumberAllDifferentDigits(int num) {
```
Нельзя руками заносить все значения, а если их будет 100 тыс, необходимо переписать через алгоритм

6ECCMEPTHIoIIJ replied:
Исправил. 
Изначально был написан алгоритм с циклом, но чуть позже пришел к выводу, что для 4-хзначного числа перебрать 9 вариантов вручную не такая проблема (потому что их немного), при этому такой вариант намного эффективнее с точки зрения времени исполнения.

6ECCMEPTHIoIIJ replied:
однако, если бы вариантов было больше, такая идея в голову бы не пришла

vasilmik commented on src/test/java/edu/hw1/Task8Test.java:
```diff
@@ -0,0 +1,260 @@
+
+    static @NotNull Stream<DeskTestArgs> nonCapturedDeskProviderFactory() {
+        return Stream.of(
+            new DeskTestArgs(new boolean[][] {
```
Сделайте нормальный тест как в задаче

6ECCMEPTHIoIIJ replied:
Добавил тесты из примера к заданию. 
Не совсем понял, что не так с моими тестами (кроме того, что true false невозможно читать). Там проверяются варианты на пустое поле и поле, где все кони не могут друг на друга напасть, а также все варианты на взаимное расположение коней (их всего 4). На каждый из 4 вариантов проверяются 5 вариантов их расположения относительно доски: в центре доски, в одном из 4-х углов

vasilmik commented on src/main/java/edu/hw1/Task8.java:
```diff
@@ -0,0 +1,38 @@
+            throw new NullPointerException();
+        }
+
+        return !isKnightOverlapped(desk, -KNIGHT_LONG_MOVE_DISTANCE, KNIGHT_SHORT_MOVE_DISTANCE)
```
Проверьте эти варианты

6ECCMEPTHIoIIJ replied:
Не могли бы вы пояснить, что именно мне необходимо проверить?

6ECCMEPTHIoIIJ replied:
В комментарии к Task8Test расписал, какие именно варианты я проверяю

6ECCMEPTHIoIIJ committed:
Task1, Task3, Task4, Task8 exception type changed.
Task2 fixed bug for Integer.MIN_VALUE. Negative numbers become illegal. Added test.
Task6 replaced constants with cycle.
Task8 added tests from example
